### PR TITLE
[Serialization] Serialize source file information to '.swiftsourceinfo' file

### PR DESCRIPTION
**Pull request by rintaro**
Serialize module source file informations to `.swiftsourceinfo` file.
* File path
* Interface hash (including type members)
* Modified time
* Size

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/Serialization/ModuleFile.cpp [hunk 1]
```diff
@@ -959,6 +959,36 @@ Optional<CommentInfo> ModuleFile::getCommentForDecl(const Decl *D) const {
...
+    llvm::function_ref<void(const BasicSourceFileInfo &)> callback) const {
+  if (Core->SourceFileListData.empty())
+    return;
+  assert(!Core->SourceLocsTextData.empty());
+
+  auto *Cursor = Core->SourceFileListData.bytes_begin();
+  auto *End = Core->SourceFileListData.bytes_end();
+  while (Cursor < End) {
+    auto fileID = endian::readNext<uint32_t, little, unaligned>(Cursor);
+    auto fpStr = StringRef{reinterpret_cast<const char *>(Cursor),
```
Comment (benlangmuir) [inline lib/Serialization/ModuleFile.cpp:972]: Heh, this took me a second to understand. Maybe worth a comment that this is a fixed-length string?

File: lib/AST/RawComment.cpp [hunk 2]
```diff
@@ -238,3 +239,27 @@ CharSourceRange RawComment::getCharSourceRange() {
...
+    return true;
+
+  FilePath = filename;
+  LastModified = stat->getLastModificationTime();
+  FileSize = stat->getSize();
+
+  if (SF->hasInterfaceHash()) {
+    InterfaceHash = SF->getInterfaceHashIncludingTypeMembers();
+  } else {
+    // FIXME: Parse the file with EnableInterfaceHash option.
```
Comment (benlangmuir) [inline lib/AST/RawComment.cpp:260]: When would this happen?

Comment (rintaro) [reply lib/AST/RawComment.cpp:260]: Whole module optimization. Interface hash is generated only when the file is explicitly specified `-primary-file`.

File: lib/Serialization/SerializeDoc.cpp [hunk 3]
```diff
@@ -783,6 +783,56 @@ static void emitBasicLocsRecord(llvm::BitstreamWriter &Out,
...
+      // Make 'FilePath' absolute for serialization;
+      SmallString<128> absolutePath = info.FilePath;
+      llvm::sys::fs::make_absolute(absolutePath);
+
+      // Don't emit duplicated files.
+      if (!seenFilenames.insert(info.FilePath).second)
+        return;
+
+      auto fileID = FWriter.getTextOffset(absolutePath);
+      auto timestamp = std::chrono::duration_cast<std::chrono::seconds>(info.LastModified.time_since_epoch()).count();
```
[the remark this answers is not in the document]
Comment (akyrtzi) [reply lib/Serialization/SerializeDoc.cpp]: Yes, there's no reason to lose accuracy here, even a single `uint64_t` is sufficient to keep nanosecond count.

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Review by akyrtzi — APPROVED** [review]

**Pull request merged**